### DISCUSSION TITLE
Stuck with the (psc)prescaler calculations

### DISCUSSION THREAD

**Discussion by NitinSaxenait** [japaric/discovery#369] [Q&A] [answered]
Hi, I am new to Embedded and Rust too. I am working on the stm32f3 discovery board and completed the 8 chapters. Now I am working with the chapter clock & timers. I read the whole chapter, registers we are using, and how we are changing our way to use delay. I have searched about the counters and Prescaler(The purpose of the Prescaler is to allow the timer to be clocked at the rate a user desires).
1. So we are first making the timer6 of apb1 clock enable. 
2. Next we are setting our counter as open pulse mode.
3. then we are giving the timer a psc value to make the delay as needed.
4. After that we are using the auto-reload register which provides the end value to the counter.
5. After crossing the end value an event occurs and a uif bit is set which means our timer is completed. 
All these points are clear to me but I am not able to catch the calculation of psc value and how the counting is done.

   
```
// Configure the prescaler to have the counter operate at 1 KHz
    tim6.psc.write(|w| w.psc().bits(psc));
```
I'm going to let you figure out the value of the prescaler, psc. Remember that the frequency of the counter is apb1 / (psc + 1) and that apb1 is 8 MHz.
This calculation. Can someone please help me I have already search youtube but not find a simple explanation?

**Comment by adamgreig** [comment, marked as answer]
It sounds like you've almost got it!

The timer is driven at 8MHz by the core, which means each "tick" (in other words, every count it makes) takes 1/8000000=125 nanoseconds. But, we'd like it to count in milliseconds instead, which means it needs to run at 1kHz.

Since the 8MHz input is fixed, we must use the prescaler to divide the incoming 8MHz frequency down to the 1kHz frequency we'd like it to run at. The output frequency is freq_out = APB1 / (PSC + 1), and APB1 is 8MHz, so we need to solve the equation 1000 = 8000000 / (PSC + 1) to find the value for PSC, 7999.

**Reply by NitinSaxenait** [reply]
Thanks @adamgreig Now I can continue further with peace
